docs/design/request-compression-trait.md

```diff
@@ -0,0 +1,546 @@
+    val byteArrayStream = ByteArrayOutputStream()
+    val gzipStream = GZIPOutputStream(byteArrayStream)
+
+    content?.let { gzipStream.use { it.write(content) } }
+
+    return byteArrayStream.toByteArray()
+}
+```
+
+Gzip JVM streaming compression implementation:
+
+```kotlin
+/**
+
+Plan on writing code to convert SourceContent & ChannelContent to  
+`java.io.OutputStream` to use `GZIPOutputStream` in similar fashion as above.
+`GZIPOutputStream` takes in a `java.io.OutputStream` which `ByteArrayOutputStream` is.
+
+Could also use function to convert stream to byte array but chances are will run out
+of memory doing so. Body will probably be large if using streaming trait on it.
+
+If that code to convert the types is not possible for some reason then will have to 
+hand write gzip implementation.
+
+---------------------------------------------------
+
+Also looked at how flexible checksums deals with this problem but it's not really transferable :/
+
+**/
+```
+
+Allowing user supplied compression algorithms is possible if it's something that wants to be done. There are a couple of issues with this feature that aren't impossible to overcome but certainly pose a challenge that might not be worth it. Some of those include:
+
+1. Server side would be unable to decompress a random compression algorithm provided by a user. There might be no way around this unless a user is implementing a compression algorithm that the SDK would not support at the time. Then there is some niche use case for this.
+
+2. Allowing a user supplied compression algorithm that adjusts the "level" of compression is a good idea but in practice the server side might not be able to handle this functionality and this could break things for the customer.
+
+3. Customers might want to use their own implementations of compressing algorithms but this seems unnecessary.
+
+That level of client configurability comes with some caveats, and a decision was made to not pursue this feature. Regardless of that this can always be added later if it starts to seem necessary/required.
+
+## Implementation
+The implementation/integration of the feature is broken down into sections to be able to dive into detail as well as clarity.
+
+### Middleware
+We will be using an interceptor to modify the request before checksums are calculated for it. Middleware is how we will be registering the interceptor. This is how this one will work:
+
+* A kotlin integration will look at a service model
+
+* Determine if the service has an operation with the `requestCompression` trait
+* If so middleware is registered  for the service
+* The middleware will look at each operation in the service individually
+* The middleware will then determine if an operation has the `requestCompression` trait
+* If so then if will add an interceptor to the operation
+
+More detail on interceptors can be found in this same directory under [interceptors.md](interceptors.md)
```

Comment:
nit: don't refer to directories

`More detail on interceptors can be found in [the design document.](interceptors.md)`